src/modm/driver/temperature/max31865.hpp

```diff
@@ -0,0 +1,249 @@
+
+	struct FaultThreshold
+	{
+		using RtdResistance_t = uint16_t;
+	};
+
+	static constexpr uint16_t
+	rtdfaultthresh(const FaultThreshold::RtdResistance_t res)
+	{
+		return static_cast<uint16_t>(res << 1);
+	};
+
+	enum class Register : uint8_t
+	{
+		ReadConfiguration = 0x00,
+		ReadRtdMsb = 0x01,
+		ReadRtdLsb = 0x02,
+		ReadHighFaultThresholdMsb = 0x03,
+		ReadHighFaultThresholdLsb = 0x04,
+		ReadLowFaultThresholdMsb = 0x05,
+		ReadLowFaultThresholdLsb = 0x06,
+		ReadFaultStatus = 0x07,
+
+		WriteConfiguration = 0x80,
+		WriteHighFaultThresholdMsb = 0x83,
+		WriteHighFaultThresholdLsb = 0x84,
+		WriteLowFaultThresholdMsb = 0x85,
+		WriteLowFaultThresholdLsb = 0x86,
+	};
+
+	struct modm_packed Data
+	{
+		/// @return measure resistance in ohm
+		constexpr float
+		getResistance() const
+		{
+			const uint16_t adccode = data >> 1;
+			return adccode / 32768.f * pt.Rref;
+		}
+
+		/// @return fast temperature in degrees celsius, about 0.3 degrees error between 0 and 100
+		/// degrees celsius
+		constexpr float
+		getTemperatureFast() const
+		{
+			return (getResistance() - pt.R0) / pt.alpha / 100.f;
+		}
+
+		/// @return slow but accurate temperature in degrees celsius
+		constexpr double
+		getTemperaturePrecise() const
+		{
+			const double res = getResistance();
+			double T = getTemperatureFast();
+			const double c0 = T <= 0 ? pt.c : 0;
+
+			// Do some fixed number of newton steps for root finding
+			// on Callendar Van Dusen equation:
+			// R = R0*(1+a*T+b*T*T+c*(T-100)*T*T*T)
+			// Newton seems to need double precision to achieve 1.e-10 residual?!
+			for (int i = 0; i < 10; i++)
+			{
+				const double R = double{pt.R0} * (1 + (pt.a * T) + (pt.b * T * T) +
+												  (c0 * (T - 100) * T * T * T)) -
+								 res;
+				const double Rdash =
+					double{pt.R0} * (pt.a + (2 * pt.b * T) + c0 * (((4 * T) - 300) * T * T));
+				T -= R / Rdash;
+				if (std::abs(R) <= 1.e-10) { break; }
+			}
+
+			return T;
+		}
+
+		uint16_t data;
+	};
+};  // struct max31865
+
+/**
+ * @tparam SpiMaster
+ * @tparam Cs
+ *
+ * @author  Henrik Hose
+ * @ingroup modm_driver_max31865
+ */
+template<typename SpiMaster, typename Cs, Pt pt = pt100>
+class Max31865 : public max31865<pt>,
+				 public modm::SpiDevice<SpiMaster>,
+				 protected modm::NestedResumable<3>
+{
+public:
+	using typename max31865<pt>::Fault;
+	using typename max31865<pt>::Fault_t;
+	using typename max31865<pt>::Config;
+	using typename max31865<pt>::Config_t;
+	using typename max31865<pt>::FaultDetectionWrite;
+	using typename max31865<pt>::FaultDetectionWrite_t;
+	using typename max31865<pt>::FaultDetectionRead;
+	using typename max31865<pt>::FaultDetectionRead_t;
+	using typename max31865<pt>::Rejection;
+	using typename max31865<pt>::Rejection_t;
+	using typename max31865<pt>::FaultThreshold;
+	using typename max31865<pt>::rtdfaultthresh;
+	using typename max31865<pt>::Register;
+	using typename max31865<pt>::Data;
```

Comment:
Does anyone know a nicer way to do this?
(I don't understand the reason why members of a templated base class are hidden in the derived class.)

Reply:
Isn't the original purpose of this base class pattern that the definitions are not dependent on the template arguments? It does not make much sense that your register definitions are different types if you change the sensor resistance.

Reply:
Move the template arguments to the inner `Data` class and make `max31865` a regular class.

Reply:
Fair point. I pushed a commit that moves the `[...]::Data` struct completely outside the `max31865` struct.
Not sure if that really better/nicer that just moving the template argument to the `Data` sub-struct...

Reply:
I would have put `Data` into `max31865` for consistency. It's like that in most other drivers. There is no other technical reason.